Steer to Zero feature

mogorman opened the pull request:
iniital work to add steer to zero interceptor.  goes almost to zero
while panda work is figured out

mogorman committed:
Steer to Zero feature

iniital work to add steer to zero interceptor.  goes almost to zero
while panda work is figured out

mogorman committed:
add cache time

mogorman committed:
one day ill get this right

mogorman committed:
Update settings.cc

mogorman committed:
Update carcontroller.py

mogorman committed:
whitespace issue

mogorman committed:
trying this

mogorman committed:
too late

mogorman committed:
fixes everyhing

j-vanetten commented on selfdrive/car/chrysler/carcontroller.py:
```diff
@@ -109,7 +109,7 @@ def update(self, enabled, CS, actuators, pcm_cancel_cmd, hud_alert, gas_resume_s
         self.hud_count += 1
 
-    new_msg = create_lkas_command(self.packer, int(apply_steer), self.gone_fast_yet, frame)
+    new_msg = create_lkas_command(self.packer, int(apply_steer), (self.gone_fast_yet and enabled), frame)
```
This is going to need some testing as I'm not sure what effect it's going to be on un-modded vehicles.

j-vanetten replied:
I checked this and this causes a lkas fault for me.  If you you need this change, you'll need to wrap it in the new flag.

mogorman replied:
i removed it from pr, in the other one sorry leaked in here.

j-vanetten commented on selfdrive/car/chrysler/carstate.py:
```diff
@@ -28,6 +28,7 @@ def __init__(self, CP):
 
   def update(self, cp, cp_cam):
     speed_adjust_ratio = self.cachedParams.get_float('jvePilot.settings.speedAdjustRatio', 5000)
+    steer_zero_enabled = self.cachedParams.get('jvePilot.settings.enableSteerToZero', 5000)
```
Make this a boolean right away.
```suggestion
    steer_zero_enabled = self.cachedParams.get('jvePilot.settings.enableSteerToZero', 5000) == "1"
```

mogorman replied:
in other pr

j-vanetten commented on selfdrive/car/chrysler/carstate.py:
```diff
@@ -74,7 +75,7 @@ def update(self, cp, cp_cam):
     ret.steeringPressed = abs(ret.steeringTorque) > STEER_THRESHOLD
     steer_state = cp.vl["EPS_STATUS"]["LKAS_STATE"]
-    ret.steerError = steer_state == 4 or (steer_state == 0 and ret.vEgo > self.CP.minSteerSpeed)
+    ret.steerError = steer_state == 4 or (steer_state == 0 and (ret.vEgo > self.CP.minSteerSpeed and steer_zero_enabled == "0"))
```
once you make `steer_zero_enabled` a boolean, it will read better here as:
```suggestion
    ret.steerError = steer_state == 4 or (steer_state == 0 and ret.vEgo > self.CP.minSteerSpeed and not steer_zero_enabled)
```

mogorman replied:
yup

j-vanetten commented on selfdrive/car/chrysler/interface.py:
```diff
@@ -20,6 +20,7 @@ def compute_gb(accel, speed):
   def get_params(candidate, fingerprint=gen_empty_fingerprint(), car_fw=None):
     speed_adjust_ratio = cachedParams.get_float('jvePilot.settings.speedAdjustRatio', 5000)
     inverse_speed_adjust_ratio = 2 - speed_adjust_ratio
+    steer_zero_enabled = cachedParams.get('jvePilot.settings.enableSteerToZero', 5000)
```
```suggestion
    steer_zero_enabled = cachedParams.get('jvePilot.settings.enableSteerToZero', 5000) == "1"
```

j-vanetten commented on selfdrive/car/chrysler/interface.py:
```diff
@@ -50,7 +51,10 @@ def get_params(candidate, fingerprint=gen_empty_fingerprint(), car_fw=None):
+      if steer_zero_enabled == "1":
+        ret.minSteerSpeed = 0
+      else:
+        ret.minSteerSpeed = 17.5 * inverse_speed_adjust_ratio  # m/s 17 on the way up, 13 on the way down once engaged.
```
```suggestion
      # m/s 17 on the way up, 13 on the way down once engaged.
      ret.minSteerSpeed = 0 if steer_zero_enabled else 17.5 * inverse_speed_adjust_ratio  
```

j-vanetten commented on selfdrive/ui/qt/offroad/settings.cc:
```diff
@@ -97,6 +97,12 @@ JvePilotTogglesPanel::JvePilotTogglesPanel(QWidget *parent) : QWidget(parent) {
+                                  "Enable Steer to Zero",
+                                  "When enabled, jvePilot will take advantage of eps intercept hardware and allow steering to 0mph.",
+                                  "../assets/offroad/icon_speed_limit.png",
+                                  this));
```
We should possible think about making steer speed more adjustable.  This would allow other to try and lower their models min steer speed to see if it works on their model.  In your case, you would enter -17.5 to get to 0, but a new model Pacifica user could try slowly going into the negative to see how much they could push it.  There may be uses out there with models that can go lower but we don't know.

I also don't really like this as a top level toggle as most people don't have this hardware.  If we do make it a number value, we could put it under the misc settings.

Open for a debate here

mogorman replied:
i think those are all good call outs. closing this for now while i can test the changes.

mogorman committed:
Feature WP Steer to Zero Mod

This branch adds support for taking advantage of a white panda to
intercept messages being sent. It enables the car to steer to 0mph
removing the 9mph and 39mph lockouts

mogorman committed:
Merge branch 'feature_wp_to_submit' into feature_wp

mogorman committed:
changes requested by j-vanetten

renamed flag, added minSteerSpeed variable, moved configuration.

mogorman committed:
Merge branch 'feature_wp_to_submit' into feature_wp
